src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,205 @@
+                current = current.next;
+                currentIndex++;
+            }
+            if (current.next == null) {
+                Node<T> newNode = new Node<>(current, value, null);
+                current.next = newNode;
+                tail = newNode;
+            } else {
+                Node<T> newNode = new Node<>(current, value, current.next);
+                current.next.prev = newNode;
+                current.next = newNode;
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
+        if (index < 0 || index >= size()) {
+            throw new IndexOutOfBoundsException("Indeks jest poza zakresem listy: " + index);
+        }
+
+        Node<T> current = head;
+        int currentIndex = 0;
+
+        while (current != null && currentIndex < index) {
+            current = current.next;
+            currentIndex++;
+        }
+        if (current != null && currentIndex == index) {
+            return current.value;
+        }
         return null;
     }
 
     @Override
     public T set(T value, int index) {
+        if (index < 0 || index >= size()) {
+            throw new IndexOutOfBoundsException("Indeks jest poza zakresem listy: " + index);
+        }
+
+        Node<T> current = head;
+        int currentIndex = 0;
+
+        while (current != null && currentIndex < index) {
+            current = current.next;
+            currentIndex++;
+        }
+
+        if (current != null && currentIndex == index) {
+            T oldValue = current.value;
+            current.value = value;
+            return oldValue;
+        }
+
         return null;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size()) {
+            throw new IndexOutOfBoundsException("Invalid index: " + index);
+        }
+
+        Node<T> nodeToRemove = getNodeAtIndex(index);
+        T valueToRemove = nodeToRemove.value;
+        removeNode(nodeToRemove);
+        return valueToRemove;
     }
 
-    @Override
     public boolean remove(T object) {
+        if (object == null) {
+            return removeNull();
+        } else {
+            return removeNonNull(object);
+        }
+    }
+
+    private boolean removeNull() {
+        Node<T> current = head;
+
+        while (current != null) {
+            if (current.value == null) {
+                removeNode(current);
+                return true;
+            }
+            current = current.next;
+        }
+
+        return false;
+    }
+
+    private boolean removeNonNull(T object) {
+        Node<T> current = head;
+
+        while (current != null) {
+            if (object.equals(current.value)) {
+                removeNode(current);
+                return true;
+            }
+            current = current.next;
+        }
```

Comment:
code duplication
merge this logic and put it into remove, you just need to combine the condition in `if`
it should be as in Objects.equals (but write it out yourself since you can't use it in this task)
`a == b || a != null && a.equals(b)`